pennylane/devices/qubit/simulate.py

```diff
@@ -237,7 +256,295 @@ def simulate(
     tensor([0.68117888, 0.        , 0.31882112, 0.        ], requires_grad=True))
 
     """
-    state, is_state_batched = get_final_state(circuit, debugger=debugger, interface=interface)
+    has_mcm = has_mid_circuit_measurements(circuit)
+    has_shots = circuit.shots.total_shots is not None
+    if has_mcm and has_shots:
+        return simulate_native_mcm(
+            circuit, rng=rng, prng_key=prng_key, debugger=debugger, interface=interface
+        )
+    state, is_state_batched, _ = get_final_state(circuit, debugger=debugger, interface=interface)
     if state_cache is not None:
         state_cache[circuit.hash] = state
     return measure_final_state(circuit, state, is_state_batched, rng=rng, prng_key=prng_key)
+
+
+# pylint: disable=too-many-arguments
+def simulate_native_mcm(
+    circuit: qml.tape.QuantumScript,
+    rng=None,
+    prng_key=None,
+    debugger=None,
+    interface=None,
+) -> Result:
+    """Simulate a single quantum script with native mid-circuit measurements.
+
+    Args:
+        circuit (QuantumTape): The single circuit to simulate
+        rng (Union[None, int, array_like[int], SeedSequence, BitGenerator, Generator]): A
+            seed-like parameter matching that of ``seed`` for ``numpy.random.default_rng``.
+            If no value is provided, a default RNG will be used.
+        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. This is
+            the key to the JAX pseudo random number generator. If None, a random key will be
+            generated. Only for simulation using JAX.
+        debugger (_Debugger): The debugger to use
+        interface (str): The machine learning interface to create the initial state with
+        state_cache=None (Optional[dict]): A dictionary mapping the hash of a circuit to the pre-rotated state. Used to pass the state between forward passes and vjp calculations.
+
+    Returns:
+        tuple(TensorLike): The results of the simulation
+    """
+    if circuit.shots.has_partitioned_shots:
+        results = []
+        for s in circuit.shots:
+            aux_circuit = circuit.copy()
+            aux_circuit._shots = qml.measurements.Shots(s)
+            results.append(simulate(aux_circuit, rng, prng_key, debugger, interface))
+        return tuple(results)
+    aux_circuit = init_auxiliary_circuit(circuit)
+    all_shot_meas, mcm_values_dict = simulate_one_shot_native_mcm(
+        aux_circuit, rng, prng_key, debugger, interface
+    )
+    list_mcm_values_dict = [mcm_values_dict]
+    for _ in range(circuit.shots.total_shots - 1):
+        one_shot_meas, mcm_values_dict = simulate_one_shot_native_mcm(
+            aux_circuit, rng, prng_key, debugger, interface
+        )
+        all_shot_meas = accumulate_native_mcm(aux_circuit, all_shot_meas, one_shot_meas)
+        list_mcm_values_dict.append(mcm_values_dict)
+    return parse_native_mid_circuit_measurements(circuit, all_shot_meas, list_mcm_values_dict)
```

Comment:
The one thing I'm having difficulty following is how the final measurements are computed. I would imagine that all measurement process (except those applied to an mcm result) need to be computed on a set of "final" samples produced during each shot. Could you point me to how these are collected and then processed?